### PR TITLE
[recurbate.com] Add recurbate extractor

### PR DESCRIPTION
**Pull request by mrscrapy**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

This is a new extractor to support the NSFW website Recurbate.com, by default the website support 1 free download, so a premium subscription is required

Fixes #

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### CODE REVIEW

**Comment by Wado729** [comment]
I updated both files on my system but when I try recurbate, I get the following error 

"ERROR: [generic] None: Unable to download webpage: HTTP Error 403: Forbidden (caused by <HTTPError 403: 'Forbidden'>); please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U"

I am using stable@2023.03.04

**Comment by mrscrapy** [comment]
> I updated both files on my system but when I try recurbate, I get the following error
> 
> "ERROR: [generic] None: Unable to download webpage: HTTP Error 403: Forbidden (caused by <HTTPError 403: 'Forbidden'>); please report this issue on https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using yt-dlp -U"
> 
> I am using stable@2023.03.04

Are you sure you have patched yt-dlp correctly? 
I'm asking because your error should mention `[Recurbate]` rather than `[generic]`

You can use the checkout command to checkout the correct PR `gh pr checkout 6297`  on your local machine.

Also, do you have a paid membership with Recurbate?
It is required to make this work.

File: yt_dlp/extractor/recurbate.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        'skip': 'Website require membership.',
+    }]
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        try:
+            webpage = self._download_webpage(url, video_id)
+        except ExtractorError as e:
+            if e.cause.code == 403:
+                self.raise_login_required(method='cookies')
```
Comment (bashonly) [inline yt_dlp/extractor/recurbate.py:26]: `import urllib.error` above and:
```suggestion
            if isinstance(e.cause, urllib.error.HTTPError) and e.cause.code == 403:
                self.raise_login_required()
            raise
```

File: yt_dlp/extractor/recurbate.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        title = self._html_extract_title(webpage, 'title')
```
Comment (bashonly) [inline yt_dlp/extractor/recurbate.py:27]: inline in return

File: yt_dlp/extractor/recurbate.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        token = self._html_search_regex(r'data-token="([^"]+)"', webpage, 'token')
+
+        video_url = f'https://recurbate.com/api/get.php?video={video_id}&token={token}'
+        video_webpage = self._download_webpage(video_url, video_id)
+
+        entries = self._parse_html5_media_entries(video_url, video_webpage, video_id)
+        if not entries:
+            if 'shall_signin' in video_webpage[:20]:
+                self.raise_login_required(method='cookies')
```
Comment (bashonly) [inline yt_dlp/extractor/recurbate.py:36]: is this still necessary if unauthenticated users will be getting 403 from the initial webpage request?

File: yt_dlp/extractor/recurbate.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+            raise ExtractorError('No media links found')
```
Comment (bashonly) [inline yt_dlp/extractor/recurbate.py:37]:
```suggestion
            self.raise_no_formats('No media links found')
```

**Comment by mrscrapy** [comment]
I think they implemented additional cookies validations and now I consistently get 403 while authenticated.
It will probably need some other work

File: yt_dlp/extractor/recurbate.py [hunk 5]
```diff
@@ -25,7 +25,7 @@ def _real_extract(self, url):
             webpage = self._download_webpage(url, video_id)
         except ExtractorError as e:
             if isinstance(e.cause, urllib.error.HTTPError) and e.cause.code == 403:
-                self.raise_login_required()
+                self.raise_login_required(msg='This video is only available for registered users; Set your authenticated browser user agent via the --user-agent parameter.', method='cookies')
```
Comment (mrscrapy) [inline yt_dlp/extractor/recurbate.py:28]: I don't know if this message is clear enough, but the --user-agent parameter is required to match the one set during the browser session.

Is there also an automated way to do that?

Comment (mrscrapy) [reply yt_dlp/extractor/recurbate.py:28]: > > Is there also an automated way to do that?
> 
> No

Well then, nothing to add.

The extractor works with the `--user-agent` parameter

File: yt_dlp/extractor/recurbate.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
+import urllib.error
+
+from .common import InfoExtractor
+from ..utils import ExtractorError, merge_dicts
+
+
+class RecurbateIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?recurbate\.com/play\.php\?video=(?P<id>[0-9]+)'
```
Comment (pukkandan) [inline yt_dlp/extractor/recurbate.py:8]:
```suggestion
    _VALID_URL = r'https?://(?:www\.)?recurbate\.com/play\.php\?video=(?P<id>\d+)'
```

File: yt_dlp/extractor/recurbate.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+        title = self._html_extract_title(webpage, 'title')
+        token = self._html_search_regex(r'data-token="([^"]+)"', webpage, 'token')
+
+        video_url = f'https://recurbate.com/api/get.php?video={video_id}&token={token}'
+        video_webpage = self._download_webpage(video_url, video_id)
+
+        entries = self._parse_html5_media_entries(video_url, video_webpage, video_id)
+        return merge_dicts({
+            'id': video_id,
+            'title': title,
```
Comment (pukkandan) [inline yt_dlp/extractor/recurbate.py:39]:
```suggestion
        token = self._html_search_regex(r'data-token="([^"]+)"', webpage, 'token')
        video_url = f'https://recurbate.com/api/get.php?video={video_id}&token={token}'
        
        video_webpage = self._download_webpage(video_url, video_id)
        entries = self._parse_html5_media_entries(video_url, video_webpage, video_id)
        return merge_dicts({
            'id': video_id,
            'title': self._html_extract_title(webpage, 'title'),
```

**Review by pukkandan — APPROVED** [review]

**Comment by pukkandan** [comment]
Pls test nothing's broken.

**Pull request merged**